File: paryty-v1.0/cluster/internal/models/agent.go
package models

import (
	"time"
)

// AgentStatus represents the current status of an agent.
type AgentStatus string

const (
	AgentStatusOnline     AgentStatus = "online"
	AgentStatusOffline    AgentStatus = "offline"
	AgentStatusConnecting AgentStatus = "connecting"
	AgentStatusError      AgentStatus = "error"
)

// AgentInfo contains information about a registered agent.
type AgentInfo struct {
	ID            string            `json:"id" db:"id"`
	Hostname      string            `json:"hostname" db:"hostname"`
	IPAddress     string            `json:"ip_address" db:"ip_address"`
	OS            string            `json:"os" db:"os"`
	Arch          string            `json:"arch" db:"arch"`
	AgentVersion  string            `json:"agent_version" db:"agent_version"`
	Labels        map[string]string `json:"labels" db:"labels"`
	Status        AgentStatus       `json:"status" db:"status"`
	RegisteredAt  time.Time         `json:"registered_at" db:"registered_at"`
	LastHeartbeat time.Time         `json:"last_heartbeat" db:"last_heartbeat"`
	LastError     string            `json:"last_error,omitempty" db:"last_error"`
	ConfigHash    string            `json:"config_hash" db:"config_hash"`
}

// AgentRegistration is the request sent by an agent to register with the cluster.
type AgentRegistration struct {
	AgentID      string            `json:"agent_id" db:"agent_id"`
	Hostname     string            `json:"hostname" db:"hostname"`
	IPAddress    string            `json:"ip_address" db:"ip_address"`
	OS           string            `json:"os" db:"os"`
	Arch         string            `json:"arch" db:"arch"`
	AgentVersion string            `json:"agent_version" db:"agent_version"`
	Labels       map[string]string `json:"labels" db:"labels"`
	Capabilities []string          `json:"capabilities" db:"capabilities"`
}

// AgentConfig is the configuration pushed to an agent by the cluster.
type AgentConfig struct {
	AgentID            string            `json:"agent_id" db:"agent_id"`
	ConfigVersion      string            `json:"config_version" db:"config_version"`
	CollectionInterval time.Duration     `json:"collection_interval" db:"collection_interval"`
	EnabledCollectors  []string          `json:"enabled_collectors" db:"enabled_collectors"`
	SamplingRate       float64           `json:"sampling_rate" db:"sampling_rate"`
	CompressionEnabled bool              `json:"compression_enabled" db:"compression_enabled"`
	FlowControl        FlowControlConfig `json:"flow_control" db:"flow_control"`
}

// FlowControlConfig contains flow control settings for an agent.
type FlowControlConfig struct {
	MaxQueueSize   int     `json:"max_queue_size" db:"max_queue_size"`
	MaxBatchSize   int     `json:"max_batch_size" db:"max_batch_size"`
	SamplingRate   float64 `json:"sampling_rate" db:"sampling_rate"`
	BackpressureOn bool    `json:"backpressure_on" db:"backpressure_on"`
}

// HealthReport is a health report sent by an agent.
type HealthReport struct {
	AgentID     string            `json:"agent_id" db:"agent_id"`
	Status      HealthStatus      `json:"status" db:"status"`
	Uptime      time.Duration     `json:"uptime" db:"uptime"`
	MemoryUsage float64           `json:"memory_usage" db:"memory_usage"`
	CPUUsage    float64           `json:"cpu_usage" db:"cpu_usage"`
	Components  map[string]string `json:"components" db:"components"`
	Timestamp   time.Time         `json:"timestamp" db:"timestamp"`
}

File: paryty-v1.0/frontend/src/types/agent.ts
// Agent domain types — mirror the cluster's Go models (snake_case wire format).
// Source of truth: cluster/internal/models/agent.go and topology.go.

import type { Labels, Timestamp } from './common';

/** Connection lifecycle state of an agent. */
export type AgentStatus = 'online' | 'offline' | 'connecting' | 'error';

/** Overall health classification shared by agents and topology nodes. */
export type HealthStatus = 'healthy' | 'degraded' | 'unhealthy' | 'unknown';

/** Edge agent lifecycle status (Dual Reality) */
export type EdgeAgentLifecycleStatus =
  | 'unconfigured' | 'active' | 'lost' | 'rogue' | 'retired' | 'blacklisted'
  | 'pending' | 'deployed' | 'inactive'; // backward compat

/**
 * Registered agent metadata.
 *
 * Field names match the cluster JSON contract exactly (`GET /api/v1/agents`,
 * `GET /api/v1/agents/:id`).
 */
export interface AgentInfo {
  id: string;
  hostname: string;
  ip_address: string;
  os: string;
  arch: string;
  agent_version: string;
  labels: Labels;
  status: AgentStatus;
  registered_at: Timestamp;
  last_heartbeat: Timestamp;
  last_error?: string;
  config_hash: string;
}

/**
 * Health report for a single agent (`GET /api/v1/agents/:id/health`).
 *
 * `uptime` is a Go `time.Duration` serialized as nanoseconds.
 */
export interface HealthReport {
  agent_id: string;
  status: HealthStatus;
  uptime: number;
  memory_usage: number;
  cpu_usage: number;
  components: Record<string, string>;
  timestamp: Timestamp;
}

/** Agent management info from `GET /api/v1/agents/all`. */
export interface AgentManagementInfo {
  agent_id: string;
  name: string;
  hostname: string;
  status: string; // pending | deployed | inactive | unconfigured | active | lost | rogue | retired | blacklisted
  os: string;
  arch: string;
  cloud_provider: string;
  location: string;
  assigned_twin: string | null;
  first_seen: string;
  last_seen: string;
  /** Timestamp when the agent was paired with a cluster agent. */
  paired_at: string | null;
  /** Timestamp when the agent was retired. */
  retired_at: string | null;
  /** Timestamp when the agent was blacklisted. */
  blacklisted_at: string | null;
  /** Reason for blacklisting, if applicable. */
  blacklist_reason: string | null;
  /** Identity token for agent authentication. */
  identity_token: string | null;
}

/** Pairing status response from backend */
export interface AgentPairingStatus {
  agent_id: string;
  edge_status: EdgeAgentLifecycleStatus;
  cluster_agent_id: string | null;
  cluster_agent_name: string | null;
  cluster_agent_status: string | null;
  is_paired: boolean;
  paired_at: string | null;
  os: string;
  arch: string;
  hostname: string;
  retired_at: string | null;
  blacklisted_at: string | null;
  blacklist_reason: string | null;
}

/** Response from `POST /api/v1/agents`. */
export interface CreateAgentResponse {
  agent_id: string;
  name: string;
  status: string;
}
